### PR TITLE
Add out-of-tree Pyodide builds in CI for `scikit-image`

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
## Description

This pull request is meant to supersede #7265 based on @lagru's suggestion, and it adds a CI job to build and test `scikit-image` in a Pyodide virtual environment through WASM wheels. Following this, it would be possible to include JupyterLite notebooks in the documentation in order to run `scikit-image`'s code snippets which come with docstring-based examples.

It is to be highlighted that this was completed last month for NumPy: (https://github.com/numpy/numpy/pull/25894) and for PyWavelets – an optional dependency for `scikit-image`: (https://github.com/PyWavelets/pywt/pull/701), all thanks to @rgommers for the help.

I was working on this last month, and I have rebased my branch to account for the latest changes upstream.

cc @stefanv from discussions on the previous PR.

## Checklist

- [x] A descriptive but concise pull request title
- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- [ ] A gallery example in `./doc/examples` for new features
- [x] [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

We use [changelist](https://github.com/scientific-python/changelist) to
compile each pull request into an item of the release notes. Please refer to
the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes)
and [past release notes](https://scikit-image.org/docs/stable/release_notes/index.html)
for guidance and examples.

```release-note
...
```

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
For context: a lot of the `.npy` files were replaced with the compressed `.npz` file format in https://github.com/scikit-image/scikit-image/pull/7350/commits/0dc74319b2dcbe3cab07d2f706a851246e6d27d9 – please see https://github.com/pyodide/pyodide/issues/4541 for more information. This fix has not been released yet, but I am happy to drop the commit once we do have a release.

**Comment by agriyakhetarpal** [comment]
Some progress: adding a JS script to run the `pytest` test suite works, and all the tests proceed to run till 100%. I will now work to resolve the failures and errors in the test cases.

**Comment by agriyakhetarpal** [comment]
Some progress: adding a JS script to run the `pytest` test suite has made it work, and all the tests proceed to run till 100% completion. I will now work to resolve the failures and errors faced in the test cases.

**Comment by agriyakhetarpal** [comment]
Looks like all tests are now passing! This is ready for review now, @lagru and @rgommers. I am happy to rewrite history here if needed, or have this can be squash merged – either way works for me.

As soon as this is merged, I can follow up with a PR that publishes these wheels weekly (based on the current cadence) to the Anaconda.org repository or any other PyPI-like indices. Doing that should be much more trivial in comparison to this PR :) I can also add a corresponding job in the `test_nightlies_on_main.yml` workflow if that is required, but I don't think we have nightly WASM wheels for a lot of repositories right now, so this can wait for a while – I would prefer to do that in a separate PR anyway!

File: skimage/_shared/testing.py
Comment (agriyakhetarpal) [inline skimage/_shared/testing.py:361]: I have shifted the threading import to inside the function – this function is not a part of the public API offered by `scikit-image`, but if needed I can make it a top-level import again and skip it on WASM.

File: skimage/feature/tests/test_util.py
Comment (agriyakhetarpal) [inline skimage/feature/tests/test_util.py:73]: Even with the `MPLBACKEND` environment variable set inside the workflow, it didn't work and the tests for Matplotlib were failing – I would suggest keeping this change here (and in other places).

File: skimage/io/tests/test_io.py
Comment (agriyakhetarpal) [inline skimage/io/tests/test_io.py:51]: I am unsure of this – could not get it to pass inside Node. Maybe it's running in a web worker already and cannot access a server? Also, `pytest-localserver` isn't available in Pyodide for some reason (even though it is a pure Python package).

File: tools/emscripten/patches/0001-fix-npy-file-access.patch
Comment (agriyakhetarpal) [inline tools/emscripten/patches/0001-fix-npy-file-access.patch:1]: The entire patch isn't needed, we require just the changes to `library_fs.js`, `library_nodefs.js`, and `library_syscall.js`. I didn't want to break the patch, so this is copied verbatim from upstream.

**Review by agriyakhetarpal — COMMENTED** [review]
Some comments to facilitate an easier review, below:

File: skimage/feature/tests/test_util.py
Comment (rgommers) [reply skimage/feature/tests/test_util.py:73]: I won't have as much context as @lagru here, but for what it's worth: in SciPy we set the backend to Agg for testing, and it's my understanding that that in general is a robust strategy.

**Review by rgommers — APPROVED** [review]
This looks quite good to me. The CI job config looks good, as do the logs. Applying the needed patch for `.npy` files is a good approach.

The changes to test files all seem reasonable; the `import threading` one is the only one that may require a change, but you already commented on that so the maintainers can decide on that one.

**Comment by agriyakhetarpal** [comment]
Pyodide had a recent patch release to fix failures on the `pyodide build` command, I updated it in d02105b4fcbe41005a5fc4655323a9d32b324333 and merged the latest changes from the `main` branch (haven't rebased so as not to spoil the review above, but that can always be done later as needed). The Emscripten toolchain version has not changed with this release, so the ABI remains stable.

**Comment by lagru** [comment]
Sorry for not keeping up with this @agriyakhetarpal. :pray: I plan to review this by the weekend.

**Comment by agriyakhetarpal** [comment]
Thanks, @lagru! No worries at all. I was merging `main` a few times over recently to ensure that the jobs are not failing with upstream changes. Plus, it would only ease the review, I imagine.

File: tools/emscripten/pytest_scikit_image_pyodide.js
Comment (lagru) [inline tools/emscripten/pytest_scikit_image_pyodide.js:7]: Could you update our [license](https://github.com/scikit-image/scikit-image/blob/main/LICENSE.txt) accordingly with

```
Files: tools/emscripten/pytest_scikit_image_pyodide.js
Copyright: 2007-2024 The scikit-learn developers
License: BSD-3-Clause
```

File: tools/emscripten/pytest_scikit_image_pyodide.js
Comment (lagru) [inline tools/emscripten/pytest_scikit_image_pyodide.js:33]: Ah, I was wondering how the build wheel was included in the pyodide venv.

File: tools/emscripten/pytest_scikit_image_pyodide.js
Comment (lagru) [inline tools/emscripten/pytest_scikit_image_pyodide.js:41]: I don't particularly like that installing these extra test dependencies is tucked away here. 

Do you think it would be feasible to try installing the dependencies in `requirements/test.yml` and just ignore deps that don't work? Alternatively, we could add an extra `requirements/pyodide_test.txt`.

File: tools/emscripten/patches/0001-fix-npy-file-access.patch
Comment (lagru) [reply tools/emscripten/patches/0001-fix-npy-file-access.patch:1]: That's fine. :+1:

File: skimage/feature/tests/test_util.py
Comment (lagru) [reply skimage/feature/tests/test_util.py:73]: I don't really follow why `use('Agg')` fixes things and `MPLBACKEND="Agg"` doesn't. E.g. for https://github.com/scikit-image/scikit-image/pull/7350/commits/be525a44f4620ee423dbd47e521c87b0bd4c3fd8, [pytest's error log](https://github.com/scikit-image/scikit-image/actions/runs/8360772772/job/22887205529#step:9:1875) already indicates that `Agg` is used.

This is before you try to [use Agg programmatically](https://github.com/scikit-image/scikit-image/pull/7350/commits/f466193dea59ce92d50ce643f9a0c70a4f7b9386)

**Review by lagru — COMMENTED** [review]
Thank you @agriyakhetarpal! I'm impressed by the workarounds you figured out by yourself without knowing our library intimately. :D

File: skimage/feature/tests/test_util.py
Comment (agriyakhetarpal) [reply skimage/feature/tests/test_util.py:73]: I don't understand it enough either (perhaps Node.js is not properly letting environment variables pass to the Python interpreter). But sure, let me try to revert this change in a separate commit, again, at the end of addressing the review and we can see if things break in an isolated state. Either way, keeping just one of these options would be better.

File: skimage/io/tests/test_io.py
Comment (agriyakhetarpal) [reply skimage/io/tests/test_io.py:51]: That's great. Let's skip them, then :)

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:30]: > Right now this is only building and testing and not uploading the wheel somewhere, correct?
>
> It's fine if we keep that out of scope for this PR but I'd still like to talk about how we might include the wheel in our [doc building workflow](https://github.com/scikit-image/scikit-image/blob/7ae4fd8833391a3fa6f28b12a53a177699518764/.github/workflows/build_docs.yml)?

Yes, that is correct. The doc builds workflow(s) aspect is still something I am experimenting with right now, but it has been completed for SciPy (https://github.com/scipy/scipy/pull/20019) and PyWavelets (https://github.com/PyWavelets/pywt/pull/728). The SciPy one is disabled right now, but the PyWavelets one is available and can be accessed on the PyWavelets documentation website at any of the examples under https://pywavelets.readthedocs.io/en/latest/ref/index.html. The issue we have at hand here is how to provide these wheels pre-installed into a JupyterLite-compatible kernel through either a custom Pyodide distribution or through any other mechanism, which I am actively working towards.

> We might also upload it as part of [nightly wheel building workflow](https://github.com/scikit-image/scikit-image/blob/main/.github/workflows/nightly_wheel_build.yml?rgh-link-date=2024-04-13T11%3A43%3A07Z) similar to [numpy](https://github.com/numpy/numpy/blob/ab7649fe2ed8f0f0260322d66631b8dfab57deff/.github/workflows/emscripten.yml#L116-L126).
> 
> In case we decide tackle this, may we ping you for support? :D

Absolutely! I would be more than happy to make the changes myself in a follow-up PR, considering that this would be beneficial for the documentation builds mentioned above – the plan will be to use these wheels, after they have been built, but only in the latest versions of the docs (the stable versions may use a pinned wheel from the Pyodide CDN), however, this is preliminary and I suspect things will change. I hope I can trouble you for a review on another PR soon after this one ;)

File: tools/emscripten/pytest_scikit_image_pyodide.js
Comment (agriyakhetarpal) [reply tools/emscripten/pytest_scikit_image_pyodide.js:7]: Thanks for the heads up – I fixed this in 19159a4412309fc9310aa93458565a0ec035cc1d.

File: tools/emscripten/pytest_scikit_image_pyodide.js
Comment (agriyakhetarpal) [reply tools/emscripten/pytest_scikit_image_pyodide.js:41]: You're right, we could do better with less repetition – I can indeed add a `requirements.txt` file, but we still don't have the necessary functionality to install from such files in `micropip` yet: https://github.com/pyodide/micropip/issues/6, i.e., the file will remain redundant for a while.

I couldn't find a`requirements/test.yml` file as you suggested, so I assume you meant the `requirements/test.txt` file – out of these:

```txt
# Generated via tools/generate_requirements.py and pre-commit hook.
# Do not edit this file; modify pyproject.toml instead.
asv
numpydoc>=1.7
pooch>=1.6.0
pytest>=7.0
pytest-cov>=2.11.0
pytest-localserver
pytest-faulthandler
pytest-doctestplus
```

dependencies, `pytest` and `pytest-cov` are likely the only ones that would work right now, and `asv` and `pooch` serve different uses that are not of use here I guess. The rest of the plugins make more sense to be running in a non-Pyodide environment and outside of the regular tests.

File: skimage/_shared/testing.py
Comment (agriyakhetarpal) [reply skimage/_shared/testing.py:239]: Thanks, that will indeed be better. Addressed in 0b1c933ca36e3f12810dd7ed27f870457a76e8df.

File: skimage/feature/_basic_features.py
Comment (agriyakhetarpal) [reply skimage/feature/_basic_features.py:95]: I think that would be nice to add (though I worry about the added complexity). I'll add this class and associated functions in the same file.

Comment (agriyakhetarpal) [reply skimage/feature/_basic_features.py:95]: Okay, after a few hours of reading the docs and some tries locally, I think I have a solution with [`contextlib.ExitStack`](https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack) for now, so something like this should work:

```python
from contextlib import ExitStack, AbstractContextManager

# this is the same one as you defined above ⬆️
class _FallbackThreadPoolExecutor(AbstractContextManager):
    def __init__(self, max_workers):
        self.max_workers = max_workers

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        pass

    def map(self, fn, iterables):
        return map(fn, iterables)

# and then we can perform a conditional entry into the context manager here
threading_available = True
if (sys.platform == "emscripten") or (platform.machine() in ["wasm32", "wasm64"]):
    threading_available = False

if threading_available:
    executor_class = ThreadPoolExecutor
else:
    executor_class = _FallbackThreadPoolExecutor

# the exitstack can manage the closing of the context manager that gets created
# with stack.enter_context() here. not sure if it can be reduced further
with ExitStack() as stack:
    if threading_available:
        ex = stack.enter_context(ThreadPoolExecutor(max_workers=num_workers))
    else:
        ex = stack.enter_context(_FallbackThreadPoolExecutor(max_workers=num_workers))
    out_sigmas = list(
        ex.map(
            lambda s: _singlescale_basic_features_singlechannel(
                img, s, intensity=intensity, edges=edges, texture=texture
            ),
            sigmas,
        )
    )
features = itertools.chain.from_iterable(out_sigmas)
return features
```

This isn't the best way IMO, but if there is a better idea or if you have suggestions on this approach, I am happy to incorporate them. I haven't committed this change yet because I feel it is a (little) less readable in comparison to guarding the `threading` import and then proceeding with a non-threaded implementation without managing contexts.

File: skimage/feature/tests/test_util.py
Comment (agriyakhetarpal) [reply skimage/feature/tests/test_util.py:73]: I have done a bit of reading about this as someone who does not have much JavaScript experience, and I think what is happening here is that Pyodide runs Python code within the browser environment, isolated from the Node.js process that spawns the Pyodide interpreter – therefore, it doesn't directly inherit environment variables from Node.js.
It could be tricky to debug further but I'm happy to do so, so I'll remove the `MPLBACKEND` environment variable from the workflow file for now, given that the programmatic approach with `matplotlib.use` will be beneficial for not only these Pyodide-specific tests, but for the general test suite on regular supported platforms too.

File: .github/workflows/emscripten.yml
Comment (lagru) [reply .github/workflows/emscripten.yml:30]: Perfect. Then let's get to that after merging this PR. :D

File: tools/emscripten/pytest_scikit_image_pyodide.js
Comment (lagru) [reply tools/emscripten/pytest_scikit_image_pyodide.js:7]: Huh, the commit isn't shown as being part of this PR and also doesn't appear in the diff. Did you push that commit somewhere else accidentally?

File: tools/emscripten/pytest_scikit_image_pyodide.js
Comment (lagru) [reply tools/emscripten/pytest_scikit_image_pyodide.js:41]: Okay, then let's keep streamlining this for a later day. We can't run the full test suite anyway yet.

File: skimage/feature/_basic_features.py
Comment (lagru) [reply skimage/feature/_basic_features.py:95]: This seems a lot more complicated than what I had in mind. Perhaps you misunderstood? Let me investigate this locally and see if I can push something simpler..

File: tools/emscripten/pytest_scikit_image_pyodide.js
Comment (agriyakhetarpal) [reply tools/emscripten/pytest_scikit_image_pyodide.js:7]: That's strange... I am able to see 19159a4412309fc9310aa93458565a0ec035cc1d (re-added the commit) below:

image

Here is the link to the commit that displays the diff from the PR's context: [`19159a4` (#7350)](https://github.com/scikit-image/scikit-image/pull/7350/commits/19159a4412309fc9310aa93458565a0ec035cc1d), maybe this helps?

File: skimage/_shared/testing.py
Comment (lagru) [reply skimage/_shared/testing.py]: Oh, I meant to return even earlier and skip the definition of `inner`, see https://github.com/scikit-image/scikit-image/pull/7350/commits/e94b1cd15082f5ff2afc0ef0dcabae9cb6f100a2.

File: skimage/feature/_basic_features.py
Comment (lagru) [reply skimage/feature/_basic_features.py]: See https://github.com/scikit-image/scikit-image/pull/7350/commits/cdd9830199da49824ff5dd843ca18c1b0da3c0b9.

Comment (agriyakhetarpal) [reply skimage/feature/_basic_features.py]: Yes, I think so, too. An investigation from your end would be helpful :)

File: tools/emscripten/pytest_scikit_image_pyodide.js
Comment (lagru) [reply tools/emscripten/pytest_scikit_image_pyodide.js:7]: I can see it now. Probably me or my browser being confused. :thinking: Sorry.

**Comment by agriyakhetarpal** [comment]
Hi, @lagru! It is almost going to be a month (in a few hours) since this was last reviewed – I hope it is okay if I can ping you again so that we can bring this to completion? If it helps, I can either wait for a while until Pyodide releases version 0.26.0 (https://github.com/pyodide/pyodide/issues/4713, https://github.com/pyodide/pyodide/discussions/4751) and rebase the changes to work with the newer Emscripten version's ABI and Python version bump, but even without that, this is an incremental process, 0.25.1 is perfectly fine and I can do the upgrade in a later PR as well.

Once we get this in, I can follow up as mentioned in https://github.com/scikit-image/scikit-image/pull/7350#issuecomment-2009846294 and https://github.com/scikit-image/scikit-image/pull/7350#discussion_r1565727339 to get nightly uploads for the wheels going, and then possibly get started on the interactive documentation part.

**Comment by agriyakhetarpal** [comment]
I see https://github.com/scikit-image/scikit-image/issues/7282 also has a few pending tasks that are probably more pressing, but Pyodide 0.26.0 will still have NumPy v1.26.4; perhaps @rgommers will have more context on getting NumPy 2.0 in-tree support, but since builds for NumPy pass out-of-tree quite nicely, it shouldn't be an issue to update it in-tree when things are ready and we see a non-release-candidate release.

The question there would be whether that would require a 0.26.1 release or 0.27.0, or whether it should be included in 0.26.0 and Pyodide should wait until releasing it, since building with NumPy 2 could be a big endeavour for many downstream packages despite the prevalent communication processes that have been put in place around it over the past many months.

**Comment by rgommers** [comment]
Please don't worry about NumPy 2.0, it's not relevant in this context. Scikit-image (and every other core scientific Python package) will support 1.26 and 2.0 in parallel, so Pyodide will just contain one of those and it doesn't matter which one.

**Comment by lagru** [comment]
Hey @agriyakhetarpal, sorry for the silence on my (and our) part. I've just had other more pressing things to attend but I fully intend to review and merge this! Thank you again for the work so far. :pray:

This PR is in my backlog and I'll try to get to it as soon as possible. Also, if there are any time critical considerations from your side, please let me know. :)

**Comment by agriyakhetarpal** [comment]
No time-critical considerations from my side! I just merged `main` to check if all tests still pass and to see if there's something new that fails – my task is to ease the review for you.

**Comment by agriyakhetarpal** [comment]
a6754d5273835735223c93c95e2edb6746bc2a24 bumps the Pyodide version to the newly out [`0.26.0`](https://github.com/pyodide/pyodide/releases/tag/0.26.0). In addition, there is [preliminary support for `cibuildwheel`](https://cibuildwheel.pypa.io/en/latest/setup/#pyodide-webassembly-builds-experimental) to build for Pyodide as well – happy to try that out if needed!

**Comment by lagru** [comment]
Thanks for staying on top of this. I've added `mymindstorm/setup-emsdk@*` to the set of allowed actions. ;)

**Comment by agriyakhetarpal** [comment]
The Pyodide build didn't run for this set of PR checks, so maybe it needs to be triggered again. I just merged `main` for doing so. Could you please approve the workflows, @lagru?

File: .github/workflows/emscripten.yml
Comment (lagru) [inline .github/workflows/emscripten.yml:16]:
```suggestion
    branches:
      - main
      - maintenance/**
```
It seems like this is close to ready. So we don't forget before merging.

Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:16]: Yes, thanks. Let me add this after fixing the failing tests.

Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml]: Addressed in 7b33abd87f56a9ac6e926bff8ac0c74d7e81f79e.

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [inline .github/workflows/emscripten.yml:33]:
```suggestion
    if: github.repository == 'scikit-image/scikit-image'
```

**Comment by agriyakhetarpal** [comment]
The test warnings/failures have been fixed and are now passing, plus I cleaned up the workflow added. This should be ready for review again. I can rebase and rewrite history if needed, or let this be squash merged, both options would work. I'll follow these changes up by working towards https://github.com/scikit-image/scikit-image/pull/7350#discussion_r1563973109.

File: .github/workflows/emscripten.yml
Comment (lagru) [inline .github/workflows/emscripten.yml:11]: Is there a reason to keep the pull request trigger? Since we have `if: github.repository == 'scikit-image/scikit-image'` enabled below, a pull request targeting either `main` or `maintenance` would still be skipped.

File: .github/workflows/emscripten.yml
Comment (lagru) [inline .github/workflows/emscripten.yml:19]: Just curious, at what tool is this variable aimed at?

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:11]: Ideally, we'll want that the out-of-tree builds will run on every PR similar to standard PR checks, in order to catch errors early/during PRs. I'm not sure I follow, though. The condition `if: github.repository == 'scikit-image/scikit-image'` will stop the workflow from running on pull requests on forks, but it will allow running on PRs to `scikit-image` regardless of where the base branch of the PR is (whether it's coming from a fork, or a collaborator's branch for those who have write access to the repo). A build triggered by a pull request to `scikit-image` should not get skipped when targeting either `main` or `maintenance`, as far as I understand.

File: skimage/io/tests/test_mpl_imshow.py
Comment (lagru) [inline skimage/io/tests/test_mpl_imshow.py:39]: So this is a warning that has to be addressed upstream?

**Review by lagru — APPROVED** [review]
I've added some minor comments or questions but this looks ready to merge IMO. Thank you so much for leading this effort @agriyakhetarpal and for the impressively fast iteration cycle on your side! :ok_hand:

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:19]: I have been referencing this from https://learn.scientific-python.org/development/guides/tasks/#nox through which GHA can display coloured outputs for tests – it works well for regular Python/`pytest` test suites. I don't think it's working in this Node.js test runner, however. I am happy with either removing it or keeping it for other jobs. :)

File: .github/workflows/emscripten.yml
Comment (lagru) [reply .github/workflows/emscripten.yml:11]: Ah, I think I was confused. Then, is there any reason not to give it the same permissions as the `test.yml` workflow using 
```yml
on: [push, pull_request, merge_group]
```

And there's also no reason to keep the `if: github.repository == 'scikit-image/scikit-image'` right?

File: skimage/io/tests/test_mpl_imshow.py
Comment (agriyakhetarpal) [reply skimage/io/tests/test_mpl_imshow.py:39]: Yes – this is most likely coming from https://github.com/python/cpython/issues/101588. Pyodide won't be able to address this until it bumps to CPython 3.13, which will happen in the next minor release (0.27.X) – the CPython version is coupled with the Pyodide version (and as is the Emscripten toolchain version).

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:11]: Yes, there's no reason as such, it's mostly because I have been implementing the workflow in the same way across projects, and I guess skipping the workflow on forks saves time with an instant notification :) Let me remove this and synchronise the `test.yml` workflow and this one in the next commit.

Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml]: Addressed in 81d1eabea53d3e22ea1b71ed5987e431fa6a0348.

Comment (lagru) [reply .github/workflows/emscripten.yml]: I don't follow, in what way does this save time? :thinking:

Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml]: and in 67aa8ee36d2f3c40a82c5f82e9539741063122a0.

Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml]: > I don't follow, in what way does this save time? 🤔

Ah, I worded that a bit wrong. If one syncs their `main` branch with the upstream `main` branch, and that triggers the workflow runs, it would skip the workflow instantly. Skipping workflows would also be useful for developers who are trying to test out something on their forks so that they don't have to disable the workflows manually. Though, I would agree that this is a niche use case and my reasoning is built upon other projects that skip workflows.

File: .github/workflows/emscripten.yml
Comment (stefanv) [reply .github/workflows/emscripten.yml:20]: Ideally, we'd store the built wheel as an artifact, grab it from the docs build, and upload it as part of the docs. @agriyakhetarpal it should be possible to mpip or similar from there?

File: .github/workflows/emscripten.yml
Comment (stefanv) [inline .github/workflows/emscripten.yml:53]: This approach is totally fine, just wondering about the advantage of the manual vs cibuildwheel approach?

File: skimage/feature/_basic_features.py
Comment (stefanv) [inline skimage/feature/_basic_features.py:16]: Is it necessary to define `__init__`?

**Review by stefanv — APPROVED** [review]
This is great, thanks @agriyakhetarpal! I left some minor comments.

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:20]: Yes. This is something that I am trying to get done for PyWavelets first through the [provided artifacts](https://jupyterlite.readthedocs.io/en/latest/howto/pyodide/wheels.html), but this is still in-progress work. For installing from `pip`, JupyterLite provides [abstractions over `micropip`](https://jupyterlite.readthedocs.io/en/latest/howto/pyodide/packages.html#installing-packages-at-runtime) to install packages (it won't work for Anaconda.org packages of the [lack of CORS headers](https://github.com/pyodide/micropip/issues/101) and a host of other reasons, but the Anaconda index at least serves as a nice place to host these wheels going forward). I would suggest looking at the Anaconda upload and then serving them through the docs in separate PRs so that they are easier to review for the maintainers.

File: .github/workflows/emscripten.yml
Comment (agriyakhetarpal) [reply .github/workflows/emscripten.yml:53]: I am happy to implement `cibuildwheel`, but it was very recently merged (https://github.com/pypa/cibuildwheel/pull/1456) and there hasn't been a release yet. I did try it for NumPy (https://github.com/numpy/numpy/pull/26570), pandas (https://github.com/pandas-dev/pandas/pull/58647), and PyWavelets (https://github.com/PyWavelets/pywt/pull/744), and all of them work – so if it would be okay to use using `pypa/cibuildwheel@main` for now and merge, I am happy to implement the changes as necessary.

File: skimage/feature/_basic_features.py
Comment (agriyakhetarpal) [reply skimage/feature/_basic_features.py:16]: I don't think it should be needed – let me remove this.

Comment (agriyakhetarpal) [reply skimage/feature/_basic_features.py]: Addressed in 0bcf3c091c22715becb429b2e23af27c51c5d66f.

Comment (agriyakhetarpal) [reply skimage/feature/_basic_features.py]: Never mind, it does pass the number of workers as an argument, so it's needed: https://github.com/agriyakhetarpal/scikit-image/actions/runs/9385733448/job/25844607528

**Comment by lagru** [comment]
Thanks @agriyakhetarpal! 

Long-term we probably want this to run only on `main` and not for every pull request.

**Comment by agriyakhetarpal** [comment]
> Long-term we probably want this to run only on `main` and not for every pull request.

Thanks for the merge! Is there a reason why, though – I imagine there could be uncaught errors through PRs dealing with compiled code that may make it to `main`? Also, in the longer term, I would suggest moving this to a docs-building workflow so that these wheels can be built nightly and utilised as a part of interactive docs for the `latest` version (which I will be happy to implement). 

In other words, I'm going to follow up with a PR right now that uploads these wheels to Anaconda.org, as we discussed above – with that, `scikit-image` will become the third package to upload nightly WASM wheels.

**Comment by stefanv** [comment]
I guess there's a slight concern about the number of CPU cycles spun, but it's not a big deal. And it would be great to have this part of the devel docs workflow!

**Pull request merged**
